pkg/validation/app_test.go

```diff
@@ -735,6 +735,174 @@ func Test_ValidateMetadataConstraints(t *testing.T) {
 	}
 }
 
+func Test_ValidateNamespace(t *testing.T) {
+	ctx := context.Background()
+
+	tests := []struct {
+		name        string
+		obj         v1alpha1.App
+		apps        []*v1alpha1.App
+		expectedErr string
+	}{
+		{
+			name: "case 0: flawless",
+			obj: v1alpha1.App{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "kiam",
+					Namespace: "eggs2",
+				},
+				Spec: v1alpha1.AppSpec{
+					Catalog:   "giantswarm",
+					Name:      "kiam",
+					Namespace: "kube-system",
+					NamespaceConfig: v1alpha1.AppSpecNamespaceConfig{
+						Annotations: map[string]string{
+							"linkerd.io/inject": "enabled",
+						},
+					},
+					Version: "1.4.0",
+				},
+			},
+			apps: []*v1alpha1.App{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "another-kiam-1",
+						Namespace: "eggs2",
+					},
+					Spec: v1alpha1.AppSpec{
+						Catalog:   "giantswarm",
+						Name:      "kiam",
+						Namespace: "kube-system",
+						NamespaceConfig: v1alpha1.AppSpecNamespaceConfig{
+							Annotations: map[string]string{
+								"linkerd.io/inject": "enabled",
+							},
+						},
+						Version: "1.3.0-rc1",
+					},
+				},
+			},
+		},
+		{
+			name: "case 1: namespace annotation collision",
+			obj: v1alpha1.App{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "kiam",
+					Namespace: "eggs2",
+				},
+				Spec: v1alpha1.AppSpec{
+					Catalog:   "giantswarm",
+					Name:      "kiam",
+					Namespace: "kube-system",
+					NamespaceConfig: v1alpha1.AppSpecNamespaceConfig{
+						Annotations: map[string]string{
+							"linkerd.io/inject": "enabled",
+						},
+					},
+					Version: "1.4.0",
+				},
+			},
+			apps: []*v1alpha1.App{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "another-kiam-1",
+						Namespace: "eggs2",
+					},
+					Spec: v1alpha1.AppSpec{
+						Catalog:   "giantswarm",
+						Name:      "kiam",
+						Namespace: "kube-system",
+						NamespaceConfig: v1alpha1.AppSpecNamespaceConfig{
+							Annotations: map[string]string{
+								"linkerd.io/inject": "disabled",
+							},
+						},
+						Version: "1.3.0-rc1",
+					},
+				},
+			},
+			expectedErr: "app `kiam` is modifying annotation `linkerd.io/inject` in the target namespace `kube-system` from value `enabled` to `disabled`, collide to app `another-kiam-1`",
```

Comment:
```suggestion
			expectedErr: "app `kiam` annotation `linkerd.io/inject` for target namespace `kube-system` collides with value `disabled` for app `another-kiam-1`",
```

Reply:
Thanks, actually I copy the result from my unittest 😄